File: scripts/subnet.chatgpt.py
import ipaddress

def get_subnet_range(ip_cidr):
    # Create an IPv4Network object from the IP address and subnet mask
    network = ipaddress.IPv4Network(ip_cidr, strict=False)
    # Get the first and last IP addresses in the subnet
    first_ip = str(network.network_address)
    last_ip = str(network.broadcast_address)
    return (first_ip, last_ip)

def get_subnet_range_binary(ip_cidr):
    ip_address, cidr = ip_cidr.split("/")
    # Convert IP address to binary
    ip_address_binary = '.'.join([bin(int(x) + 256)[3:] for x in ip_address.split(".")])
    # Get the number of bits in the subnet mask
    cidr = int(cidr)
    # Get the subnet mask in binary
    subnet_binary = "1"*cidr + "0"*(32-cidr)
    # Divide the subnet mask into four 8-bit octets
    subnet_binary = ".".join([subnet_binary[i:i+8] for i in range(0, len(subnet_binary), 8)])
    return (ip_address_binary, subnet_binary)

def binary_to_ipv4(binary_ip):
    # Split the binary IP address into octets
    octets = binary_ip.split(".")
    # Convert each octet to decimal
    decimals = [int(octet, 2) for octet in octets]
    # Join the decimal octets with '.' to form the normal IP address
    return '.'.join(map(str, decimals))


# Example usage
ip_cidr = "192.168.56.0/21"
first_ip, last_ip = get_subnet_range(ip_cidr)
print("First IP:", first_ip)
print("Last IP:", last_ip)
ip_address_binary, subnet_binary = get_subnet_range_binary(ip_cidr)
print("IP address in binary:",ip_address_binary)
print("Subnet mask in binary:",subnet_binary)
print("IP address:",binary_to_ipv4(ip_address_binary))
print("Subnet mask:",binary_to_ipv4(subnet_binary))
binary_ip = "11000000.10101000.00111000.00000000"
normal_ip = binary_to_ipv4(binary_ip)
print("Normal IP:", normal_ip)

File: scripts/subnet_chatgpt.rs
use std::str::FromStr;

use std::net::Ipv4Addr;

fn get_subnet_range(ip_cidr: &str) -> (String, String, String) {
    let mut parts = ip_cidr.split("/");
    let ip = parts.next().unwrap();
    let cidr = parts.next().unwrap();
    let ip = Ipv4Addr::from_str(ip).unwrap();
    let cidr = cidr.parse::<u8>().unwrap();
    let mask = 32 - cidr;
    let subnet = Ipv4Addr::new(!0, !0, !0, !0 << mask);
    let last_ip = Ipv4Addr::new(
        ip.octets()[0] | !(!0 << mask),
        ip.octets()[1] | !(!0 << mask),
        ip.octets()[2] | !(!0 << mask),
        ip.octets()[3] | !(!0 << mask),
    );
    let first_ip = ip;
    (subnet.to_string(), last_ip.to_string(), first_ip.to_string())
}

fn main() {
    let ip_cidr = "192.168.56.0/21";
    let (subnet, last_ip, first_ip) = get_subnet_range(ip_cidr);
    println!("Subnet mask: {}", subnet);
    println!("Last IP: {}", last_ip);
    println!("First IP: {}", first_ip);
}
